msal/src/Microsoft.Identity.Client/Internal/Requests/DeviceCodeRequest.cs

```diff
@@ -0,0 +1,133 @@
+
+        public DeviceCodeRequest(AuthenticationRequestParameters authenticationRequestParameters, Action<DeviceCodeResult> deviceCodeResultCallback) 
+            : base(authenticationRequestParameters)
+        {
+            _deviceCodeResultCallback = deviceCodeResultCallback ?? throw new ArgumentNullException("A deviceCodeResultCallback must be provided for Device Code authentication to work properly");
+            
+            LoadFromCache = false;  // no cache lookup for token
+            SupportADFS = false;
+            StoreToCache = true;
+        }
+
+        internal override async Task PreTokenRequestAsync(CancellationToken cancellationToken)
+        {
+            await base.PreTokenRequestAsync(cancellationToken).ConfigureAwait(false);
+
+            OAuth2Client client = new OAuth2Client();
+
+            var deviceCodeScopes = new HashSet<string>();
+            deviceCodeScopes.UnionWith(AuthenticationRequestParameters.Scope);
+            deviceCodeScopes.Add(OAuth2Value.ScopeOfflineAccess);
+            deviceCodeScopes.Add(OAuth2Value.ScopeProfile);
+            deviceCodeScopes.Add(OAuth2Value.ScopeOpenId);
+
+            client.AddBodyParameter(OAuth2Parameter.ClientId, AuthenticationRequestParameters.ClientId);
+            client.AddBodyParameter(OAuth2Parameter.Scope, deviceCodeScopes.AsSingleString());
+
+            // Talked with Shiung, devicecode will be added to the discovery endpoint "soon".
+            // Fow now, the string replace is correct.
+            // TODO: We should NOT be talking to common, need to work with henrik/bogdan on why /common is being set
+            // as default for msal.
+            string deviceCodeEndpoint = AuthenticationRequestParameters.Authority.TokenEndpoint
+                .Replace("token", "devicecode")
+                .Replace("common", "organizations");
+
+            AuthenticationRequestParameters.Authority.TokenEndpoint = AuthenticationRequestParameters.Authority.TokenEndpoint.Replace("common", "organizations");
+
+            DeviceCodeResponse response = await client.ExecuteRequestAsync<DeviceCodeResponse>(
+                new Uri(deviceCodeEndpoint), 
+                HttpMethod.Post,
+                AuthenticationRequestParameters.RequestContext).ConfigureAwait(false);
+
+            // todo: fix resource/scopes here for how we want to invoke this...
+            _deviceCodeResult = response.GetResult(AuthenticationRequestParameters.ClientId, "todo resource");
+            _deviceCodeResultCallback(_deviceCodeResult);
+        }
+
+        protected override async Task SendTokenRequestAsync(CancellationToken cancellationToken)
+        {
+            TimeSpan timeRemaining = _deviceCodeResult.ExpiresOn - DateTimeOffset.UtcNow;
+
+            while (timeRemaining.TotalSeconds > 0.0)
+            {
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    throw new OperationCanceledException();

```

Comment:
>OperationCanceledException [](start = 30, length = 26)

thinking, we might want to have a Client exception which is an operation Canceled exception, though not sure about this as it will not be standard (but something for the PM when considering the programming model for the developer)